gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,695 @@
+
+      case STRING:
+        return new StringConverter(fieldName, nullable, repeated);
+
+      case ARRAY:
+        return new ArrayConverter(fieldName, nullable, schemaNode, state);
+
+      case ENUM:
+        return new EnumConverter(fieldName, nullable, schemaNode, state);
+
+      case RECORD:
+        return new RecordConverter(fieldName, nullable, schemaNode, state);
+
+      case MAP:
+        return new MapConverter(fieldName, nullable, schemaNode, state);
+
+      default:
+        throw new UnsupportedOperationException(fieldType + " is unsupported");
+    }
+  }
+
+  /**
+   * Converts a JsonElement into a supported ParquetType
+   * @author tilakpatidar
+   *
+   */
+  public static abstract class JsonElementConverter {
+    private final String name;
+    private final boolean nullable;
+    private final boolean repeated;
+
+    /**
+     * HashMap to convert {@link SourceType} to {@link PrimitiveTypeName}
+     */
+    protected final static HashMap<SourceType, PrimitiveTypeName> typeMap = new HashMap<>();
+
+    static {
+      typeMap.put(INT, INT32);
+      typeMap.put(LONG, INT64);
+      typeMap.put(FLOAT, PrimitiveTypeName.FLOAT);
+      typeMap.put(DOUBLE, PrimitiveTypeName.DOUBLE);
+      typeMap.put(BOOLEAN, PrimitiveTypeName.BOOLEAN);
+      typeMap.put(STRING, BINARY);
+    }
+
+    /**
+     * @param fieldName
+     * @param nullable
+     * @param repeated
+     */
+    public JsonElementConverter(String fieldName, boolean nullable, boolean repeated) {
+      this.name = fieldName;
+      this.nullable = nullable;
+      this.repeated = repeated;
+    }
+
+    /**
+     * Field name from schema
+     * @return
+     */
+    public String getName() {
+      return this.name;
+    }
+
+    /**
+     * is field nullable
+     * @return
+     */
+    public boolean isNullable() {
+      return this.nullable;
+    }
+
+    /**
+     * Returns a {@link PrimitiveType} schema
+     * @return
+     */
+    protected Type schema() {
+      return new PrimitiveType(repeated ? REPEATED : optionalOrRequired(), getTargetType(), getName());
+    }
+
+    /**
+     * Convert value
+     * @param value is JsonNull will return null if allowed or exception if not allowed
+     * @return Parquet safe type
+     */
+    public Object convert(JsonElement value) {
+      if (value.isJsonNull()) {
+        if (this.nullable) {
+          return null;
+        }
+        throw new RuntimeException("Field: " + getName() + " is not nullable and contains a null value");
+      }
+      return convertField(value);
+    }
+
+    /**
+     * Convert JsonElement to Parquet type
+     * @param value
+     * @return
+     */
+    abstract Object convertField(JsonElement value);
+
+    /**
+     * Parquet data type after conversion
+     * @return
+     */
+    public abstract PrimitiveTypeName getTargetType();
+
+    /**
+     * Returns the source type
+     * @return
+     */
+    public abstract SourceType getSourceType();
```

Comment:
When you pass in the `sourceType` to the constructor, the default implementation can just return it. I don't see the value of this get method, though.

Reply:
👍